File: python/grid.py
from cell import *

class grid:
    cells = []
    rows = []
    cols = []
    squares = []

    def __init__(self):
        self.cells = [cell(n) for n in range(81)]

        for i in range(9):
            self.rows.append([_find_i(i, j) for j in range(9)])
            self.cols.append([_find_i(j, i) for j in range(9)])
            self.squares.append([_find_i((i//3)*3 +j, (i//3)*3 +k) for j in range(3) for k in range(3)])

        for i, c in enumerate(self.cells):
            i_row, i_col = _find_coords(i)

            row = [_find_i(i_row, j) for j in range(9)]
            col = [_find_i(j, i_col) for j in range(9)]
            square = [_find_i((i_row//3)*3 +j, (i_col//3)*3 +k) for j in range(3) for k in range(3)]

            row.remove(i)
            col.remove(i)
            square.remove(i)
            
            c.row = [self.cells[j] for j in row]
            c.col = [self.cells[j] for j in col]
            c.square = [self.cells[j] for j in square]

    def get_cell(self, i_row, i_col):
        return self.cells[_find_i(i_row, i_col)]

    def set_cells(self, *tuples):
        for t in tuples:
            a, b, n = t
            self.cells[_find_i(a, b)].set(n)

    def parse_string(s):
        assert len(s) == 81+8
        tuples = []
        lines = s.split('.')
        for i, line in enumerate(lines):
            assert len(line) == 9
            for j, char in enumerate(line):
                if char == " ":
                    pass
                elif char in "123456789":
                    tuples.append((i, j, int(char)))
                else:
                    print("incorrent string!")
                    exit(1)
        G = grid()
        G.set_cells(*tuples)
        return G

    def __repr__(self):
        row_str = "{}{}{} {}{}{} {}{}{} | {}{}{} {}{}{} {}{}{} | {}{}{} {}{}{} {}{}{}"
        format = lambda a, n: row_str.format(*[cell.candidates[i] if i < len(cell.candidates) else " " for cell in self.cells[9*a:9*(a+1)] for i in range(3*n,3*(n+1))])
        return "\n".join([
            format(0, 0), format(0, 1), format(0, 2),
            " "*12+"|"+" "*13+"|",
            format(1, 0), format(1, 1), format(1, 2),
            " "*12+"|"+" "*13+"|",
            format(2, 0), format(2, 1), format(2, 2),
            "-"*39,
            format(3, 0), format(3, 1), format(3, 2),
            " "*12+"|"+" "*13+"|",
            format(4, 0), format(4, 1), format(4, 2),
            " "*12+"|"+" "*13+"|",
            format(5, 0), format(5, 1), format(5, 2),
            "-"*39,
            format(6, 0), format(6, 1), format(6, 2),
            " "*12+"|"+" "*13+"|",
            format(7, 0), format(7, 1), format(7, 2),
            " "*12+"|"+" "*13+"|",
            format(8, 0), format(8, 1), format(8, 2),
            ""
        ])

def _find_coords(i):
    i_row = i//9
    i_col = i%9
    return i_row, i_col

def _find_i(i_row, i_col):
    return 9*i_row + i_col

File: java/Grid.java

public class Grid {
    private Cell[] cells = new Cell[81];
    private Cell[][] rows = new Cell[9][9];
    private Cell[][] squares = new Cell[9][9];

    public Grid() {
        initializeCells();
    }

    public Grid(String repr) {
        initializeCells();
        parseRepr(repr);
    }

    private void initializeCells() {
        for (int i = 0; i < 81; i++) {
            this.cells[i] = new Cell(i);
        }

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                Cell cell = this.cells[9*i +j];

                this.rows[i][j] = cell;
                this.squares[3*(i/3) +(j/3)][3*(i%3) +(j%3)] = cell;
            }
        }

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                Cell cell = this.cells[9*i +j];
                int curr_square = 3*(i%3) + (j%3);

                Cell[] row = new Cell[8];
                Cell[] col = new Cell[8];
                Cell[] square = new Cell[8];

                int offset_row = 0;
                int offset_col = 0;
                int offset_square = 0;
                for (int k = 0; k < 8; k++) {
                    if (k == j) {
                        offset_row = 1;
                    }
                    if (k == i) {
                        offset_col = 1;
                    }
                    if (k == curr_square) {
                        offset_square = 1;
                    }

                    row[k] = this.rows[i][k+offset_row];
                    col[k] = this.rows[k+offset_col][j];
                    square[k] = this.squares[3*(i/3) +(j/3)][k+offset_square];
                }

                cell.finishInit(row, col, square);
            }
        }
    }

    @Deprecated
    public void setCell(int n, int v) {
        this.cells[n].set(v);
    }

    private void parseRepr(String repr) {
        String[] rows = repr.split("\\.");
        assert rows.length == 9;
        for (int i = 0; i < rows.length; i++) {
            String line = rows[i];
            assert line.length() == 9;

            for (int j = 0; j < 9; j++) {
                char c = line.charAt(j);
                if (c != ' ') {
                    int val = c - '0';
                    this.cells[9*i +j].set(val);
                }
            }
        }
    }

    @Override
    public String toString() {
        Character[][] values = new Character[27][27];
        for (int i = 0; i < 27; i++) {
            for (int j = 0; j < 27; j++) {
                int value = 3*(i%3) +(j%3) +1;
                values[i][j] = this.rows[i/3][j/3].probe(value)
                    ? (char)('0'+value)
                    : ' ';
            }
        }
        String formatString = "%c%c%c %c%c%c %c%c%c | %c%c%c %c%c%c %c%c%c | %c%c%c %c%c%c %c%c%c\n";
        String delim1 = "            |             |\n";
        String delim2 = "---------------------------------------\n";
        return 
            String.format(formatString, (Object [])values[0])+
            String.format(formatString, (Object [])values[1])+
            String.format(formatString, (Object [])values[2])+
            delim1+
            String.format(formatString, (Object [])values[3])+
            String.format(formatString, (Object [])values[4])+
            String.format(formatString, (Object [])values[5])+
            delim1+
            String.format(formatString, (Object [])values[6])+
            String.format(formatString, (Object [])values[7])+
            String.format(formatString, (Object [])values[8])+
            delim2+
            String.format(formatString, (Object [])values[9])+
            String.format(formatString, (Object [])values[10])+
            String.format(formatString, (Object [])values[11])+
            delim1+
            String.format(formatString, (Object [])values[12])+
            String.format(formatString, (Object [])values[13])+
            String.format(formatString, (Object [])values[14])+
            delim1+
            String.format(formatString, (Object [])values[15])+
            String.format(formatString, (Object [])values[16])+
            String.format(formatString, (Object [])values[17])+
            delim2+
            String.format(formatString, (Object [])values[18])+
            String.format(formatString, (Object [])values[19])+
            String.format(formatString, (Object [])values[20])+
            delim1+
            String.format(formatString, (Object [])values[21])+
            String.format(formatString, (Object [])values[22])+
            String.format(formatString, (Object [])values[23])+
            delim1+
            String.format(formatString, (Object [])values[24])+
            String.format(formatString, (Object [])values[25])+
            String.format(formatString, (Object [])values[26]);
    }
}

